README.md
# Arrow: Go Date Parsing

![Build Status](https://github.com/bmuller/arrow/workflows/ci/badge.svg)
[![GoDoc](https://godoc.org/github.com/bmuller/arrow?status.png)](https://pkg.go.dev/github.com/bmuller/arrow)

Arrow provides a C style format based parsing in Golang (among other helpful date/time functions).

```
Time flies like an arrow; fruit flies like a banana.
```

## Installation
Go get it:

```bash
go get github.com/bmuller/arrow
```

## Strftime Compatability
The problem with formatting times in Golang is that the [format string you give](http://golang.org/pkg/time/#Time.Format) is based on re-formatting a single date that is a pain to remember (I know it's 1/2 3:04:05 2006 -0700, but [I'm lazy](http://threevirtues.com/)).  Most languages based on C (Python, etc) use a string formatting based on [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html), which is what I think most people are familiar with.

So here's `strftime` compatability for Golang with `CFormat` and `CParse`:

```go
package main

import (
	"fmt"

	"github.com/bmuller/arrow"
)

func main() {
     // formatting
     fmt.Println("Current date: ", arrow.Now().CFormat("%Y-%m-%d %H:%M"))

     // parsing
     parsed, _ := arrow.CParse("%Y-%m-%d", "2015-06-03")
     fmt.Println("Some other date: ", parsed)
}
```

## Additional Fun
You can also utilize helpful functions to get things like the beginning of the minute, hour, day, week, month, and year.

```go
day := arrow.Now().AtBeginningOfWeek().CFormat("%Y-%m-%d")
fmt.Println("First day of week: ", day)

hour := arrow.Now().AtBeginningOfHour().CFormat("%Y-%m-%d %H:%M")
fmt.Println("First second of hour: ", hour)
```

You can also more easily sleep until specific times:

```go
// sleep until the next minute starts
arrow.SleepUntil(arrow.NextMinute())
fmt.Println(arrow.Now().CFormat("%H:%M:%S"))
```

There are also helpers to get today, yesterday, and UTC times:

```go
day := arrow.Yesterday().CFormat("%Y-%m-%d")
fmt.Println("Yesterday: ", day)

dayutc := arrow.UTC().Yesterday().CFormat("%Y-%m-%d %H:%M")
fmt.Println("Yesterday, UTC: ", dayutc)

newyork := arrow.InTimezone("America/New_York").CFormat("%H:%M:%s")
fmt.Println("Time in New York: ", newyork)
```

And for generating ranges when you need to iterate:

```go
// Print every minute from now until 24 hours from now
for _, a := range arrow.Now().UpTo(arrow.Tomorrow(), arrow.Minute) {
     fmt.Println(a.CFormat("%Y-%m-%d %H:%M:%S"))
}
```

## Running Tests
To run tests:

```
go test
```

## Reporting Issues
Please report all issues [on github](https://github.com/bmuller/arrow/issues).


CHANGELOG.md
# Changelog for v1.x.x

## v1.0.4 (2023-09-10)

### Enhancements

  * Support for UNIX-style non-zero padded notation (#3)


arrow_test.go
package arrow

import (
	"testing"
	"time"
)

func loadLocation(name string) *time.Location {
	loc, err := time.LoadLocation(name)
	if err != nil {
		return nil
	}
	return loc
}

func TestCParse(t *testing.T) {
	tests := []struct {
		dateStr  string
		timezone string
		format   string
		expected time.Time
	}{
		{
			dateStr:  "1980-06-19",
			timezone: "America/New_York",
			format:   "%Y-%m-%d",
			expected: time.Date(1980, time.June, 19, 0, 0, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-19",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d",
			expected: time.Date(1980, time.June, 19, 0, 0, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-19 9:02",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d %-I:%M",
			expected: time.Date(1980, time.June, 19, 9, 2, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-19 9:02",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d %-H:%M",
			expected: time.Date(1980, time.June, 19, 9, 2, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-19 09:02",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d %-H:%M",
			expected: time.Date(1980, time.June, 19, 9, 2, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-19 9:02",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d %H:%M",
			expected: time.Date(1980, time.June, 19, 9, 2, 0, 0, loadLocation("America/New_York")),
		},
		{
			dateStr:  "1980-6-0109:02",
			timezone: "America/New_York",
			format:   "%Y-%-m-%d%H:%M",
			expected: time.Date(1980, time.June, 1, 9, 2, 0, 0, loadLocation("America/New_York")),
		},
	}
	for _, test := range tests {
		loc, _ := time.LoadLocation(test.timezone)
		actual, err := CParseInLocation(test.format, test.dateStr, loc)
		if err != nil {
			t.Errorf("Error parsing date: %s", err)
		}
		if !actual.Time.Equal(test.expected) {
			t.Errorf("CParseInLocation(%s, %s, %s): Expected %s, got %s", test.format, test.dateStr, test.timezone, test.expected, actual)
		}
	}
}

func TestCFormat(t *testing.T) {
	tests := []struct {
		datetime time.Time
		format   string
		expected string
	}{
		{
			datetime: time.Date(1980, time.June, 19, 0, 0, 0, 0, time.UTC),
			format:   "%Y-%m-%d",
			expected: "1980-06-19",
		},
		{
			datetime: time.Date(2022, time.January, 8, 0, 0, 0, 0, time.UTC),
			format:   "%Y-%-m-%-d",
			expected: "2022-1-8",
		},
		{
			datetime: time.Date(2022, time.January, 8, 21, 2, 0, 0, time.UTC),
			format:   "%Y-%-m-%-d %-I:%-M",
			expected: "2022-1-8 9:2",
		},
		{
			datetime: time.Date(2022, time.January, 8, 4, 2, 0, 0, time.UTC),
			format:   "%Y-%-m-%-d %-H:%-M",
			expected: "2022-1-8 4:2",
		},
		{
			datetime: time.Date(2022, time.January, 8, 13, 2, 0, 0, time.UTC),
			format:   "%Y-%-m-%-d %-H:%-M",
			expected: "2022-1-8 13:2",
		},
		{
			datetime: time.Date(2006, time.August, 9, 23, 59, 11, 0, time.UTC),
			format:   "%Y-%-m-%-dT%H:%M:%S",
			expected: "2006-8-9T23:59:11",
		},
	}
	for _, test := range tests {
		a := New(test.datetime)
		actual := a.CFormat(test.format)
		if actual != test.expected {
			t.Errorf("CFormat(%s) returned %s, expected %s", test.format, actual, test.expected)
		}
	}
}


arrow.go
/*
Package arrow provides C-style date formating and parsing, along with other date goodies.

See the github project page at http://github.com/bmuller/arrow for more info.
*/
package arrow

import (
	"strconv"
	"strings"
	"time"
)

type Arrow struct {
	time.Time
}

// Like time's constants, but with Day and Week
const (
	Nanosecond  time.Duration = 1
	Microsecond               = 1000 * Nanosecond
	Millisecond               = 1000 * Microsecond
	Second                    = 1000 * Millisecond
	Minute                    = 60 * Second
	Hour                      = 60 * Minute
	Day                       = 24 * Hour
	Week                      = 7 * Day
)

func New(t time.Time) Arrow {
	return Arrow{t}
}

func UTC() Arrow {
	return New(time.Now().UTC())
}

func Unix(sec int64, nsec int64) Arrow {
	return New(time.Unix(sec, nsec))
}

func Epoch() Arrow {
	return Unix(0, 0)
}

func Now() Arrow {
	return New(time.Now())
}

func Yesterday() Arrow {
	return Now().Yesterday()
}

func Tomorrow() Arrow {
	return Now().Tomorrow()
}

func NextSecond() Arrow {
	return Now().AddSeconds(1).AtBeginningOfSecond()
}

func NextMinute() Arrow {
	return Now().AddMinutes(1).AtBeginningOfMinute()
}

func NextHour() Arrow {
	return Now().AddHours(1).AtBeginningOfHour()
}

func NextDay() Arrow {
	return Now().AddDays(1).AtBeginningOfDay()
}

func SleepUntil(t Arrow) {
	time.Sleep(t.Sub(Now()))
}

// Get the current time in the given timezone.
// The timezone parameter should correspond to a file in the IANA Time Zone database,
// such as "America/New_York".  "UTC" and "Local" are also acceptable.  If the timezone
// given isn't valid, then no change to the timezone is made.
func InTimezone(timezone string) Arrow {
	return Now().InTimezone(timezone)
}

func (a Arrow) Before(b Arrow) bool {
	return a.Time.Before(b.Time)
}

func (a Arrow) After(b Arrow) bool {
	return a.Time.After(b.Time)
}

func (a Arrow) Equal(b Arrow) bool {
	return a.Time.Equal(b.Time)
}

// Return an array of Arrow's from this one up to the given one,
// by duration.  For instance, Now().UpTo(Tomorrow(), Hour)
// will return an array of Arrow's from now until tomorrow by
// hour (inclusive of a and b).
func (a Arrow) UpTo(b Arrow, by time.Duration) []Arrow {
	var result []Arrow
	if a.After(b) {
		a, b = b, a
	}
	for a.Before(b) || a.Equal(b) {
		result = append(result, a)
		a = a.Add(by)
	}
	return result
}

func (a Arrow) Yesterday() Arrow {
	return a.AddDays(-1)
}

func (a Arrow) Tomorrow() Arrow {
	return a.AddDays(1)
}

func (a Arrow) UTC() Arrow {
	return New(a.Time.UTC())
}

func (a Arrow) Sub(b Arrow) time.Duration {
	return a.Time.Sub(b.Time)
}

// Add any duration parseable by time.ParseDuration
func (a Arrow) AddDuration(duration string) Arrow {
	if pduration, err := time.ParseDuration(duration); err == nil {
		return a.Add(pduration)
	}
	return a
}

func (a Arrow) Add(d time.Duration) Arrow {
	return New(a.Time.Add(d))
}

// The timezone parameter should correspond to a file in the IANA Time Zone database,
// such as "America/New_York".  "UTC" and "Local" are also acceptable.  If the timezone
// given isn't valid, then no change to the timezone is made.
func (a Arrow) InTimezone(timezone string) Arrow {
	if location, err := time.LoadLocation(timezone); err == nil {
		return New(a.In(location))
	}
	return a
}

func (a Arrow) AddDays(days int) Arrow {
	return New(a.AddDate(0, 0, days))
}

func (a Arrow) AddMonths(months int) Arrow {
	return New(a.AddDate(0, months, 0))
}

func (a Arrow) AddYears(years int) Arrow {
	return New(a.AddDate(years, 0, 0))
}

func (a Arrow) AddHours(hours int) Arrow {
	year, month, day := a.Time.Date()
	hour, min, sec := a.Time.Clock()
	d := time.Date(year, month, day, hour+hours, min, sec, a.Nanosecond(), a.Location())
	return New(d)
}

func (a Arrow) AddMinutes(minutes int) Arrow {
	year, month, day := a.Time.Date()
	hour, min, sec := a.Time.Clock()
	d := time.Date(year, month, day, hour, min+minutes, sec, a.Nanosecond(), a.Location())
	return New(d)
}

func (a Arrow) AddSeconds(seconds int) Arrow {
	year, month, day := a.Time.Date()
	hour, min, sec := a.Time.Clock()
	d := time.Date(year, month, day, hour, min, sec+seconds, a.Nanosecond(), a.Location())
	return New(d)
}

func (a Arrow) AtBeginningOfSecond() Arrow {
	return New(a.Truncate(Second))
}

func (a Arrow) AtBeginningOfMinute() Arrow {
	return New(a.Truncate(Minute))
}

func (a Arrow) AtBeginningOfHour() Arrow {
	return New(a.Truncate(Hour))
}

func (a Arrow) AtBeginningOfDay() Arrow {
	d := time.Duration(-a.Hour()) * Hour
	return a.AtBeginningOfHour().Add(d)
}

func (a Arrow) AtBeginningOfWeek() Arrow {
	days := time.Duration(-1*int(a.Weekday())) * Day
	return a.AtBeginningOfDay().Add(days)
}

func (a Arrow) AtBeginningOfMonth() Arrow {
	days := time.Duration(-1*int(a.Day())+1) * Day
	return a.AtBeginningOfDay().Add(days)
}

func (a Arrow) AtBeginningOfYear() Arrow {
	days := time.Duration(-1*int(a.YearDay())+1) * Day
	return a.AtBeginningOfDay().Add(days)
}

// Add any durations parseable by time.ParseDuration
func (a Arrow) AddDurations(durations ...string) Arrow {
	for _, duration := range durations {
		a = a.AddDuration(duration)
	}
	return a
}

func formatConvert(format string) string {
	// create mapping from strftime to time in Go
	strftimeMapping := map[string]string{
		"%a":  "Mon",
		"%A":  "Monday",
		"%b":  "Jan",
		"%B":  "January",
		"%c":  "", // locale not supported
		"%C":  "06",
		"%d":  "02",
		"%-d": "2",
		"%D":  "01/02/06",
		"%e":  "_2",
		"%E":  "", // modifiers not supported
		"%F":  "2006-01-02",
		"%G":  "%G", // special case, see below
		"%g":  "%g", // special case, see below
		"%h":  "Jan",
		"%H":  "15",
		"%-H": "%-H", // special case, see below
		"%I":  "03",
		"%-I": "3",
		"%j":  "%j", // special case, see below
		"%k":  "%k", // special case, see below
		"%l":  "_3",
		"%m":  "01",
		"%-m": "1",
		"%M":  "04",
		"%-M": "4",
		"%n":  "\n",
		"%O":  "", // modifiers not supported
		"%p":  "PM",
		"%P":  "pm",
		"%r":  "03:04:05 PM",
		"%R":  "15:04",
		"%s":  "%s", // special case, see below
		"%S":  "05",
		"%-S": "5",
		"%t":  "\t",
		"%T":  "15:04:05",
		"%u":  "%u", // special case, see below
		"%U":  "%U", // special case, see below
		"%V":  "%V", // special case, see below
		"%w":  "%w", // special case, see below
		"%W":  "%W", // special case, see below
		"%x":  "%x", // locale not supported
		"%X":  "%X", // locale not supported
		"%y":  "06",
		"%Y":  "2006",
		"%z":  "-0700",
		"%Z":  "MST",
		"%+":  "Mon Jan _2 15:04:05 MST 2006",
		"%%":  "%%", // special case, see below
	}

	for fmt, conv := range strftimeMapping {
		format = strings.Replace(format, fmt, conv, -1)
	}

	return format
}

// Parse the time using the same format string types as strftime
// See http://man7.org/linux/man-pages/man3/strftime.3.html for more info.
func CParse(layout, value string) (Arrow, error) {
	return CParseInLocation(layout, value, nil)
}

// Parse the time using the same format string types as strftime,
// within the given location.
// See http://man7.org/linux/man-pages/man3/strftime.3.html for more info.
func CParseInLocation(layout, value string, loc *time.Location) (Arrow, error) {
	parseFunc := func(layout string, value string, loc *time.Location) (time.Time, error) {
		if loc == nil {
			return time.Parse(layout, value)
		} else {
			return time.ParseInLocation(layout, value, loc)
		}
	}
	layout = strings.Replace(layout, "%-H", "%H", -1)
	t, e := parseFunc(formatConvert(layout), value, loc)
	return New(t), e
}

// Parse the time using the same format string types as strftime,
// within the given location (string value for timezone).
// See http://man7.org/linux/man-pages/man3/strftime.3.html for more info.
func CParseInStringLocation(layout, value, timezone string) (Arrow, error) {
	if location, err := time.LoadLocation(timezone); err == nil {
		return CParseInLocation(layout, value, location)
	} else {
		return New(time.Time{}), err
	}
}

// Format the time using the same format string types as strftime.
// See http://man7.org/linux/man-pages/man3/strftime.3.html for more info.
func (a Arrow) CFormat(format string) string {
	format = a.Format(formatConvert(format))

	year, week := a.ISOWeek()
	yearday := a.YearDay()
	weekday := a.Weekday()
	syear := strconv.Itoa(year)
	sweek := strconv.Itoa(week)
	syearday := strconv.Itoa(yearday)
	sweekday := strconv.Itoa(int(weekday))

	if a.Year() > 999 {
		format = strings.Replace(format, "%G", syear, -1)
		format = strings.Replace(format, "%g", syear[2:4], -1)
	}

	format = strings.Replace(format, "%j", syearday, -1)
	if a.Hour() < 10 {
		shour := " " + strconv.Itoa(a.Hour())
		format = strings.Replace(format, "%k", shour, -1)
	}
	format = strings.Replace(format, "%s", strconv.FormatInt(a.Unix(), 10), -1)

	if weekday == 0 {
		format = strings.Replace(format, "%u", "7", -1)
	} else {
		format = strings.Replace(format, "%u", sweekday, -1)
	}

	format = strings.Replace(format, "%-H", strconv.Itoa(a.Hour()), -1)
	format = strings.Replace(format, "%U", weekNumber(a, time.Sunday), -1)
	format = strings.Replace(format, "%U", sweek, -1)
	format = strings.Replace(format, "%w", sweekday, -1)
	format = strings.Replace(format, "%W", weekNumber(a, time.Monday), -1)
	return strings.Replace(format, "%%", "%", -1)
}

// Used for %U and %W:
// %U: The week number of the current year as a decimal number, range
// 00 to 53, starting with the first Sunday as the first day of week 01.
//
// %W: The week number of the current year as a decimal number, range
// 00 to 53, starting with the first Monday as the first day of week 01.
func weekNumber(a Arrow, firstday time.Weekday) string {
	dayone := a.AtBeginningOfYear()
	for dayone.Weekday() != time.Sunday {
		dayone = dayone.AddDays(1)
	}
	week := int(a.Sub(dayone.AddDays(-7)) / Week)
	return strconv.Itoa(week)
}
